### PR TITLE
Feature/29_LISTING

### PR DESCRIPTION
**Pull request by Symphonie234**
- Introduced a new controller named "ListingController" to manage listings.
- Implemented new code within the HomeController.
- Added multiple new migration files to handle database changes.
- Created new blade folders and placeholder index files for future use.
- Introduced a new request file to manage data validation.
- Added new model files to represent data structures.
- Included several new seeder files to populate the database.
- Introduced new CSS and JavaScript files for enhanced styling and functionality.
- Utilized TinyMCE for text editing capabilities in listing creation and editing.
- Defined new routes in the web.php file to handle listing-related actions.

### CODE REVIEW

File: app/Http/Controllers/ListingController.php [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+
+    /**
+     * Store a newly created listing in storage.
+     *
+     * @param  \Illuminate\Http\Request  $request
+     * @return \Illuminate\Http\Response
+     */
+    public function store(Request $request)
+    {
+        $listing = Listing::createListing($request);
```
Comment (ChronoDevs) [inline app/Http/Controllers/ListingController.php:54]: make error message if response is false

File: app/Http/Controllers/ListingController.php [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+
+        // Attach the selected category to the listing
+        $listing->categories()->attach($request->input('category'));
```
Comment (ChronoDevs) [inline app/Http/Controllers/ListingController.php]: add this inside Listing::createListing() model

File: app/Http/Controllers/ListingController.php [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+     * @return \Illuminate\Http\Response
+     */
+    public function update(Request $request, Listing $listing)
+    {
+        // Check if the authenticated user is the creator of the listing
+        if (Gate::denies('update-listing', $listing)) {
+            abort(403, 'Unauthorized');
+        }
+
+        $request->validate([
```
Comment (ChronoDevs) [inline app/Http/Controllers/ListingController.php]: make validation file

File: app/Http/Controllers/ListingController.php [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+            'title' => 'required|max:255',
+            'description' => 'required',
+            'tags' => 'array', // Ensure 'tags' is an array in the request
+        ]);
+
+        $listing->update([
```
Comment (ChronoDevs) [inline app/Http/Controllers/ListingController.php]: make update function in the listing model together with tags and categories and use beginTransaction()

Comment (ChronoDevs) [reply app/Http/Controllers/ListingController.php]: also make error message if response is false

File: app/Http/Controllers/ListingController.php [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+            abort(403, 'Unauthorized');
+        }
+
+        // Update the 'deleted_by' column with the ID of the user who deleted the listing
+        $listing->update([
+            'deleted_by' => Auth::id(),
+        ]);
+
+        // Soft delete the listing
+        $listing->delete();
```
Comment (ChronoDevs) [inline app/Http/Controllers/ListingController.php]: make error message if response is false

File: app/Http/Controllers/ListingController.php [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+     */
+    public function destroy(Listing $listing)
+    {
+        // Check if the authenticated user is the creator of the listing
+        if (Gate::denies('delete-listing', $listing)) {
+            abort(403, 'Unauthorized');
+        }
+
+        // Update the 'deleted_by' column with the ID of the user who deleted the listing
+        $listing->update([
```
Comment (ChronoDevs) [inline app/Http/Controllers/ListingController.php]: to override the model softdelete

Listing.php
    /**
     * Trigger eloquent events to delete.
     *
     * @return void
     */
    public function softDelete()
    {
        $this->deleted_at = $this->freshTimestamp();
        $this->deleted_by= Auth::id;
        $this->timestamps = false;
        $this->save();
    }

File: app/Models/Listing.php [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+        return $this->belongsToMany(Tag::class);
+    }
+
+    /**
+     * Create a new listing and attach tags.
+     *
+     * @param  \Illuminate\Http\Request  $request
+     * @return \App\Models\Listing
+     */
+    public static function createListing(Request $request)
```
Comment (ChronoDevs) [inline app/Models/Listing.php:44]: use beginTransaction()

**Comment by Symphonie234** [comment]
- Made error message in store and destroy functions in ListingController.
- Made a new request file named ListingUpdateRequest.
- Modified the store, update, and destroy methods in the ListingController.

**Comment by Symphonie234** [comment]
- Removed the unneccessary "@endsection" on certain blade files.
- Added "use Illuminate\Support\Facades\Auth" in the Listing model.

File: app/Http/Controllers/ListingController.php [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+        return view('listings.create', compact('tags', 'categories'));
+    }
+
+    /**
+     * Store a newly created listing in storage.
+     *
+     * @param  \Illuminate\Http\Request  $request
+     * @return \Illuminate\Http\Response
+     */
+    public function store(Request $request)
```
Comment (ChronoDevs) [inline app/Http/Controllers/ListingController.php]: use ListingStoreRequest

**Comment by Symphonie234** [comment]
- Change the parameter type of the store method from "Request" to "ListingStoreRequest"
- Adjusted the comment to that change

File: app/Http/Controllers/ListingController.php [hunk 9]
```diff
@@ -0,0 +1,142 @@
...
+     */
+    public function store(ListingStoreRequest $request)
+    {
+        // Attempt to create a new listing
+        $listing = Listing::createListing($request);
+
+        // Check if the listing was successfully created
+        if ($listing) {
+            // Attach the selected category to the listing
+            $listing->categories()->attach($request->input('category'));
```
Comment (ChronoDevs) [inline app/Http/Controllers/ListingController.php:59]: add this inside createListing function

Comment (ChronoDevs) [reply app/Http/Controllers/ListingController.php:59]: Please merge after fix. Thank you

**Pull request merged**